### PR TITLE
Mmap: fix typing of raw file descriptors

### PR DESCRIPTION
**Pull request by jiahao**
I'm not sure if this is at all the correct fix for this problem, but it does fix a failing test in `Requests.jl` which uses mmaps.

It looks like at some point raw file handles were re-typed from plain old `Int`s to `Base.Libc.RawFD`s, but several places in `Mmap` continue to emit and expect `Int`s - not updated to follow suit?

The use of `mmap` in `Requests.jl` fails on v0.4 because of what looks like an `Int` that was left out of this conversion.

``` jl
ERROR: LoadError: MethodError: `settings` has no method matching settings(::Base.Libc.RawFD, ::Bool)
Closest candidates are:
  settings(::Int64, ::Bool)
 in mmap at mmap.jl:115
 in write_file at /Users/jiahao/.julia/v0.4/Requests/src/Requests.jl:356
 in do_multipart_send at /Users/jiahao/.julia/v0.4/Requests/src/Requests.jl:443
 in send_multipart at /Users/jiahao/.julia/v0.4/Requests/src/Requests.jl:521
 in do_request at /Users/jiahao/.julia/v0.4/Requests/src/Requests.jl:567
 in post at /Users/jiahao/.julia/v0.4/Requests/src/Requests.jl:584
 in include at ./boot.jl:254
 in include_from_node1 at ./loading.jl:197
 in process_options at ./client.jl:308
 in _start at ./client.jl:411
while loading /Users/jiahao/.julia/v0.4/Requests/test/runtests.jl, in expression starting on line 139
```

I've tested these changes locally on my MBP and both the base tests as those in `Requests.jl` now pass.

These changes could use tests but I'm not sure I'm qualified to write any.

### CODE REVIEW

File: base/mmap.jl [hunk 1]
```diff
@@ -16,7 +16,7 @@ Base.isopen(::Anonymous) = true
 Base.isreadable(::Anonymous) = true
 Base.iswritable(a::Anonymous) = !a.readonly
 
-const INVALID_HANDLE_VALUE = -1
+const INVALID_HANDLE_VALUE = Libc.RawFD(-1)
```
Comment (quinnj) [inline base/mmap.jl]: Pretty sure this is going to fail on windows....

File: base/mmap.jl [hunk 2]
```diff
@@ -82,14 +82,14 @@ const FILE_MAP_EXECUTE       = UInt32(0x20)
...
-    return Int(handle)
+    return Libc.RawFD(handle)
 end
 
 settings(sh::Anonymous) = utf16(sh.name), sh.readonly, sh.create
 settings(io::IO) = convert(Ptr{Cwchar_t},C_NULL), isreadonly(io), true
 end # @windows_only
 
-# core impelementation of mmap
+# core implementation of mmap
```
Comment (quinnj) [inline base/mmap.jl]: unrelated change ;)

File: base/mmap.jl (hunk 1 above)
Comment (jiahao) [inline base/mmap.jl]: Is this even the correct diagnosis?

**Comment by quinnj** [comment]
Hmmm...I don't know about this. On the one hand, it seems like we could make it all use `RawFD`; on the other hand, I really don't see why Requests.jl is fiddling with `Base.File` or `RawFD`. It should probably just be defined on `IOStream` and be done with it. It's long overdue for an overhaul.

**Comment by jiahao** [comment]
It isn't... the first error I found was triggered by using one of the mmap constructors inside Requests.

**Comment by quinnj** [comment]
I'm talking about the excessive support for "file types" [here](https://github.com/JuliaWeb/Requests.jl/blob/a801e34dd075054ccff02e96a82c7711e51d4975/src/Requests.jl#L276). It should support `IO`, `String`, `Vector{UInt8}` and be done.

That's a little beside the point that there probably needs to be a cleaner interface between `IOStream`, `Base.File`, `Base.RawFD`, `fd()`, `open`, and `Base.__get_os_handle()` (lol).

**Comment by jiahao** [comment]
Hmm... I don't really know Requests very well.

**Comment by jiahao** [comment]
@quinnj so file descriptors on Windows are `Ptr{Void}`s? Can we force cast this to an `Int`? https://ci.appveyor.com/project/StefanKarpinski/julia/build/1.0.7166/job/w0grcil50tujrx3w#L1059

**Comment by quinnj** [comment]
See my new PR #12338; that should resolve this properly.
